### ISSUE TITLE
Regexp patterns cannot end with backslash (even if backslash-escaped)

### ISSUE THREAD

**Issue by yaauie** [elastic/logstash#9701]
An issue in the treetop grammar can cause the `regexp` to be _too_ aggressive at capturing, which can ultimately cause the pipeline compilation to fail when a pattern _ends_ with a backslash (even if that backslash is backslash-escaped).

I _believe_ I know how to fix this, but would like to discuss it to consider the risk before we tackle the compilation of the treetop grammar, since the tooling to compile the grammar does not immediately appear to be working for me and fixing it may take some effort.

***

The treetop grammar currently defines a regexp as:

~~~ treetop
  rule regexp
    ( '/' ( '\/' / !'/' . )* '/'  <LogStash::Config::AST::RegExp>)
  end
~~~

In plain English that is:
 - (1) an opening forwardslash (`'/'`)
 - (2) then a sequence of zero or more of:
   - (a) a backlash-forwardslash sequence (`'\/'`); OR
   - (b) a non-newline character that is _not_ a forwardslash (`!'/' .`)
 - (3) then a closing forwardslash (`'/'`)

_note: rule 2b uses a literal "Negative Lookahead Assertion" to prevent it from matching forward slashes_

Using this definition, in a config file the sequence of characters `/foo\\bar/` will correctly generate a Regexp that matches the literal string `foo\bar`, but the pattern `/fubar\\/`, which attempts to match the literal string `fubar\` will capture the intended (3) closing forwardslash as (2a), fail to close, and continue capturing until it either hits EOL or an unescaped forward slash from another expression, typically resulting in a pipeline compilation error.

The below table shows where the parser captures too eagerly; each parenthised group in the parse column indicates the rule that caused the pattern to match, followed by the match.

| pattern | parse | state |
|-----|-----|-----|
|`/foo/`| (1`/`)(2b`f`)(2b`o`)(2b`o`)(3`/`) | OK |
|`/foo\\bar/`| (1`/`)(2b`f`)(2b`o`)(2b`o`)(2b`\`)(2b`\`)(2b`b`)(2b`a`)(2b`r`)(3`/`) | OK |
|`/foo\/bar/`| (1`/`)(2b`f`)(2b`o`)(2b`o`)(2a`\/`)(2b`b`)(2b`a`)(2b`r`)(3`/`) | OK |
|`/fubar\\/` | (1`/`)(2b`f`)(2b`u`)(2b`b`)(2b`a`)(2b`r`)(2b`\`)(2a`\/`)... | *OPEN* | 

I believe that we can redefine the grammar for regexp to capture escape sequences better without significant risk:

~~~ diff
   rule regexp
-    ( '/' ( '\/' / !'/' . )* '/'  <LogStash::Config::AST::RegExp>)
+    ( '/' ( '\' . / !'/' . )* '/'  <LogStash::Config::AST::RegExp>)
   end
~~~

In plain English that is:
 - (1) an opening forwardslash (`'/'`)
 - (2) then a sequence of zero or more of:
   - (a) a backlash followed by any non-newline character (`'\' .`); OR
   - (b) a non-newline character that is _not_ a forwardslash (`!'/' .`)
 - (3) then a closing forwardslash (`'/'`)

This will slightly change the meaning of _how_ escape sequences get parsed (e.g., sequences that matched 2b twice before may match 2a now), but because backslashes and the character that followed them were _always_ captured and this change does nothing to _process_ the escape sequences, this fix poses no external risk.

| pattern | parse | state |
|-----|-----|-----|
|`/foo/`| (1`/`)(2b`f`)(2b`o`)(2b`o`)(3`/`) | OK |
|`/foo\\bar/`| (1`/`)(2b`f`)(2b`o`)(2b`o`)(2a`\\`)(2b`b`)(2b`a`)(2b`r`)(3`/`) | OK |
|`/foo\/bar/`| (1`/`)(2b`f`)(2b`o`)(2b`o`)(2a`\/`)(2b`b`)(2b`a`)(2b`r`)(3`/`) | OK |
|`/fubar\\/` | (1`/`)(2b`f`)(2b`u`)(2b`b`)(2b`a`)(2b`r`)(2a`\/`)(3`\/`)... | OK |

**Comment by yaauie** [comment]
(note: I believe our grammar behaves in similar ways with double- and single-quoted strings that end in backslashes, and that a similar fix may be non-breaking there too)

**Comment by HonzaKral** [comment]
I believe I am hitting this when trying to set `field_split` in `kv` filter to just `/`.

The proposed fix works *only* when `config.support_escapes` is set to `true` but if it is false then even `'\'` should be a valid string according to the docs. The setting needs to be taken into account when parsing, not after like now.

**Comment by HonzaKral** [comment]
Sorry, I meant backslash of course, it was a type, I edited my comment and fixed it as not to confuse more people, my bad!

It did understand it but then the docs for `support_escapes` say:

> `\"` becomes a literal double quotation mark. `\'` becomes a literal quotation mark.

which implies that without this setting it is not so.

**Comment by yaauie** [comment]
> which implies that without this setting it is not so

Correct, but there is more than one way for it to be "not so" :weary:

The parser has _always_ allowed users to escape a quote character _from the parser_ by prefixing it with a backslash; this merely prevents the _parser_ from identifying that quote character as a close-quote for the string it is parsing. What it hasn't done, and won't do without `config.support_escapes` set to `true`, is _process_ each escape sequence into the intended character; instead, it emits the whole escape sequence without modification.

Some plugins have relied on the above behaviour and done their own escape-sequence processing, and others have relied on the above behaviour to _avoid_ escaping inputs (e.g., when creating regular expressions).

While I'm confident that 90% of our users would benefit from having `config.support_escapes` set to `true` with their existing pipeline configurations, users who are specifying filesystem paths using the backslash-separator syntax common in Windows would be adversely affected so we cannot change the default (especially those with path-names that contain sequences like `\n` or `\t`, which would be processed into newlines and tabs).

**Comment by colinsurprenant** [comment]
@HonzaKral yeah, this is what I understand  and one of the specific problem is related to plugins that require a regex option like the kv filter. The config parser has no notion of a regex option for plugins so a regex must be passed as a string option. 

@yaauie our parser already has a notion of regex but for config expressions in conditionals, maybe we could look into introducing a regex option type for plugins? Do you think that could help?

**Comment by HonzaKral** [comment]
@colinsurprenant the option for `kv` is not a regex, just a string, a list of characters (0). Solving this for regexes won't help here.

The ultimate issue here is imho that both the parser and then the individual filters (sometimes) try and resolve the escaping. This leads to at best the need for double quotes at worst the inability to do something like specifying `/` as a string value. The incosistency where the parser always (regardless of escaping settings) parses `\"` but doesn't actually evaluate it as `"` is just adding confusion.

I believe the proper solution would be to unify the escaping rules by having the parser deal with it and hand over already unescaped string values to the plugins. It would also be possible to leave this to the plugins only but That seems less ideal to me as it would be implemented in many places instead of one.

0 - https://www.elastic.co/guide/en/logstash/current/plugins-filters-kv.html#plugins-filters-kv-field_split

**Comment by colinsurprenant** [comment]
@HonzaKral ah yes sorry, my regex comment was for options like "field_split_pattern" and for some reason I forgot your original comment was with "field_split".

**Comment by yaauie** [comment]
> @yaauie our parser already has a notion of regex but for config expressions in conditionals, maybe we could look into introducing a regex option type for plugins? Do you think that could help?
> -- @colinsurprenant in [issue comment](https://github.com/elastic/logstash/issues/9701#issuecomment-403146754)

@colinsurprenant this issue is around _literal_ regexp in the pipelines; literal _strings_, which are interpreted by a plugin to be a regexp, fall into the `config.support_escapes` camp (literal strings are also affected by a similar bug)

***

> What do you recommend moving forward?
> -- @colinsurprenant in [issue comment](https://github.com/elastic/logstash/issues/9701#issuecomment-401420973)

If you see no major flaw in my analysis of the regexp and _why_ we fail to parse strings- or regexp-ending-in-escaped-backslashes, I can push forward and get to a PR with the updated treetop grammar and generated parser.

***

> I believe the proper solution would be to unify the escaping rules by having the parser deal with it and hand over already unescaped string values to the plugins. It would also be possible to leave this to the plugins only but That seems less ideal to me as it would be implemented in many places instead of one.
> -- @HonzaKral in [issue comment](https://github.com/elastic/logstash/issues/9701#issuecomment-403149622)

Your solution is that of `config.support_escapes`, and the issue outlined in this ticket would resolve the one remaining edge-case: when attempting to end a string with a backslash.

**Comment by HonzaKral** [comment]
@yaauie even with `support_escapes` this calls for `\\\\` which is not intuitive and trips people up, that is why I would not consider it the best solution which in my mind still lies in consolidating the escape handling into the parser where it imho belongs. Of course I understand that that is a big chance and it will fix the corner case which is causing problems.

**Comment by yaauie** [comment]
I agree that `"\\\\"` isn't the most intuitive way to express a single backslash, and am open to _unambiguous_ ways to make it better.

In this case, the kv plugin really shouldn't be injecting verbatim inputs into a regexp, and should perform the necessary escaping of any inputs it receives instead of requiring the _user_ to add a layer of escaping into the current options because of this shortcoming. I've filed logstash-plugins/logstash-filter-kv#66 as a potential solution.

Additionally, @colinsurprenant's suggestion of making regexp an option type for plugins could also be helpful. I'll open up a separate design discussion around that, since there may a bit of a chicken-vs-egg situation implementing in plugins that can be installed into Logstashes that don't yet have it implemented.

**Comment by colinsurprenant** [comment]
@yaauie 
> since there may a bit of a chicken-vs-egg situation implementing in plugins that can be installed into Logstashes that don't yet have it implemented.

Right. This is off-topic here but maybe for 7.0 we could discuss how to improve plugins compatibility vs logstash versions/features. That was the original intent of the `logstash-core-plugin-api` versioning. 7.0 could be an opportunity to reboot that correctly.

**Comment by colinsurprenant** [comment]
@HonzaKral WDYT about the `field_split_char` idea in https://github.com/logstash-plugins/logstash-filter-kv/issues/66 ?

@yaauie 
> If you see no major flaw in my analysis of the regexp and why we fail to parse strings- or regexp-ending-in-escaped-backslashes, I can push forward and get to a PR with the updated treetop grammar and generated parser.

Yes, I think we should move forward.

**Comment by HonzaKral** [comment]
@colinsurprenant sounds good, though I would consider it more of a bug fix so I wouldn't mind the `field_split` arg to be fixed. After all specifying `\\\\` will will work since the regext will then just contain the backlash twice which is functionally the same

**Comment by blookot** [comment]
I'm referencing tests I'm running on dissect:
https://discuss.elastic.co/t/dissect-with-a-backslash-at-the-end-of-the-log-line/216630
any case anyone wants to contribute to the discuss thread as well :-)
Thanks!

**Comment by blookot** [comment]
PS This issue does not apply to regex!
I haven't set `config.support_escapes: true` (I think it's `false` by default). 
I'm using `\` only to escape `"` and this works perfectly well except when there is a `\` at the very end of the string.
I tried setting `config.support_escapes: true` and using a `\\` before the `"` that ends my dissect mapping but that wouldn't work either.
Thanks in advance for any update on this issue!

**Comment by HonzaKral** [comment]
@TheVastyDeep Not if you are using it for splitting, then there is a vast difference when you add `.*` at the end
